### PR TITLE
feat!: implement secure sandbox for running of unsanitized HTML

### PR DESCRIPTION
**Pull request by brotheroftux**
This introduces the following breaking changes:
1. Default HTML embedding mode now uses ShadowRoot instead of an iframe with `srcdoc` attribute
2. A new embedding mode, `isolated`, is now available to facilitate execution of custom scripts in a sandboxed environment:

```markdown
::: html [isolated]
We're counting. <span id="counter">0</span>

<script>
    const el = document.querySelector('#counter');

    window.setInterval(() => el.textContent = Number(el.textContent) + 1, 1000);
</script>
:::
```

### CODE REVIEW

File: src/utils/PromiseQueue.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    reject(thrown: Error) {
+        this.trySettle();
+        this.callbacks.reject(thrown);
+    }
+
+    get isSettled() {
+        return this.didSettle;
+    }
+
+    private trySettle() {
```
Comment (3y3) [inline src/utils/PromiseQueue.ts:55]: Зачем это?

File: src/iframe/IFrameController.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        baseElement.setAttribute('target', value);
+    }
+
+    onContentResize(callback: ContentResizeCallback): Unsubscribe {
+        this.resizeListeners.add(callback);
+
+        return () => this.resizeListeners.delete(callback);
+    }
+
+    private notifyResizeListeners() {
```
Comment (3y3) [inline src/iframe/IFrameController.ts:61]: Зачем это? У тебя резайз обсервер уже возвращает rect для entry за которым он наблюдает.

Comment (brotheroftux) [reply src/iframe/IFrameController.ts:61]: Там вербозное апи, которое подразумевает, что у обсервера может быть несколько таргетов, а ещё там какие-то гипер-вербозные ректы возвращаются, поэтому для чистоты логики я решил, что проще померять через `bCR`

File: src/utils/PromiseQueue.ts (hunk 1 above)
Comment (brotheroftux) [reply src/utils/PromiseQueue.ts:55]: Ты про весь класс или именно про этот метод?

File: src/iframe/IFrameController.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    on<E extends keyof Events>(eventName: E, eventHandler: Events[E]) {
+        this.events[eventName].add(eventHandler);
+
+        return () => this.events[eventName].delete(eventHandler);
+    }
+
+    private dispatchResize: ResizeObserverCallback = (entries) => {
+        for (const entry of entries) {
+            for (const handler of this.events?.resize) {
```
Comment (makhnatkin) [inline src/iframe/IFrameController.ts]: Why is there a cycle within a cycle? Is it ok?

File: src/iframe/index.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import {PostMessageChannel} from '../rpcAdapter/PostMessageChannel';
+import {APIPublisher} from '../rpcAdapter/APIPublisher';
+import {IFrameController} from './IFrameController';
+
+export type {Commands, Events} from './IFrameController';
+
+const $$PublisherInstanceSymbol = Symbol('$$RPCAPIPublisher');
+
+const onDOMReady = () => {
+    const controller = new IFrameController(globalThis.document.body);
```
Comment (makhnatkin) [inline src/iframe/index.ts:10]: Why do we use globalThis?

File: src/plugin/renderers/isolated.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        const processedContent = embedContentTransformFn?.(token.content) ?? token.content;
+
+        if (containerClassNames.length) {
+            token.attrSet('class', containerClassNames);
+        }
+
+        token.attrSet('src', isolatedSandboxHost);
+        token.attrSet('frameborder', '0');
```
Comment (makhnatkin) [inline src/plugin/renderers/isolated.ts:37]: lets use 
```
export const TokenAttr = {
    src: 'src',
    frameborder: 'frameborder',
    style: 'style',
} as const;
```

File: src/plugin/transform.ts [hunk 6]
```diff
@@ -35,78 +27,118 @@ type TransformOptions = {
...
+                embedContentTransformFn: (raw) => concatStylesIncludeDirectives(raw, styles),
+            });
+
+        (md as MarkdownItWithDirectives).renderer.rules[SHADOW_TOKEN_TYPE] =
+            makeShadowModeEmbedRenderRule({
+                containerClassNames: containerClasses,
+                embedContentTransformFn: (raw) => {
+                    const withStyles = concatStylesIncludeDirectives(raw, styles);
+
+                    // Note that sanitization is only enabled for `shadow` embedding mode,
```
Comment (makhnatkin) [inline src/plugin/transform.ts:124]: I suggest adding the sanitizer option for the iframe mode too, just do not set it by default

File: example/README.md [hunk 7]
```diff
@@ -64,3 +64,17 @@ After html text
...
     <p>Donec imperdiet tortor vitae ipsum gravida euismod. Donec lobortis orci erat, rutrum consequat orci aliquet non. Curabitur non dui eget orci maximus dapibus id vitae orci. Praesent sed venenatis ipsum. Suspendisse bibendum tincidunt arcu, id tempor felis consectetur non. Sed sit amet purus ultrices, tristique enim eu, efficitur mauris. Curabitur sed efficitur ligula, et varius lectus. Morbi libero purus, eleifend in vehicula eu, sagittis eu lacus. Nulla nisi ligula, mollis eu neque ornare, scelerisque rutrum enim. Duis ut volutpat neque. Vivamus gravida, felis a laoreet auctor, leo lacus mattis lacus, non auctor elit tellus ut odio. Phasellus facilisis efficitur dolor, ut fermentum mi mattis in. Suspendisse potenti. Donec tincidunt nunc eu lacus ultricies, a imperdiet est congue. Maecenas non metus non purus lobortis dignissim. Nulla libero dolor, mattis sit amet massa sed, porttitor fringilla tellus.</p>
 :::
+
+## Isolated IFrame (with script execution capabilities)
+
+Please note that you have to provide the `isolatedSandboxHost` plugin option, which should specify a **cross-origin** URL where the `iframe-runtime.html` is hosted.
```
Comment (makhnatkin) [inline example/README.md:70]: I suggest adding information on how to work with the debug server

File: src/plugin/renderers/isolated.ts (hunk 5 above)
Comment (brotheroftux) [reply src/plugin/renderers/isolated.ts:37]: Does it really make a difference, though? I actually think that would make this code less readable.

File: example/README.md (hunk 7 above)
Comment (brotheroftux) [reply example/README.md]: done

File: src/iframe/index.ts (hunk 4 above)
Comment (brotheroftux) [reply src/iframe/index.ts:10]: The proposal is already live: https://github.com/tc39/proposal-global

**Comment by makhnatkin** [comment]
lets use `feat!`
https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with--to-draw-attention-to-breaking-change

File: src/plugin/transform.ts [hunk 8]
```diff
@@ -2,30 +2,31 @@ import directivePlugin from 'markdown-it-directive';
...
     bundle: boolean;
+    isolatedSandboxHost?: string;
     sanitize?: (dirtyHtml: string) => string;
+    /**
+     * @deprecated Use the 'head' method instead.
+     */
     styles?: string | StylesObject;
+    /**
+     * @deprecated Use the 'head' method instead.
+     */
```
Comment (makhnatkin) [inline src/plugin/transform.ts:27]: lets delete @deprecated options and add info in pull request description

**Review by makhnatkin — APPROVED** [review]
Check resize plz

**Comment by makhnatkin** [comment]
- Resize delay and iframe padding are now non-configurable
- React integration now uses a different API
- Controller classes are renamed and are now using a different API via async initialize() and disposeChildren()
- Two new embedding modes were implemented: shadow, which uses a ShadowRoot instead of iframe with srcdoc attribute, and isolated which uses an iframe which should be hosted on a different origin. The latter is designed to facilitate running of potentially unsafe JS inside of HTML with no sanitization. The mode which will be used can be configured via the plugin options.

**Pull request merged**